src/main/java/org/icgc/argo/program_service/services/ProgramService.java

```diff
@@ -1,488 +1,460 @@
+    val countries = countryRepository.findAllByNameIn(program.getCountriesList());
+
+    // Add new institutions if we must
+    List<InstitutionEntity> institutions =
+        institutionRepository.findAllByNameIn(program.getInstitutionsList()); // MUTABLE
+    if (institutions.size() != program.getInstitutionsList().size()) {
+      institutions = filterAndAddInstitutions(program.getInstitutionsList());
+    }
+
+    List<ProgramCancer> programCancers = mapToList(cancers, x -> createProgramCancer(p, x).get());
+    List<ProgramPrimarySite> programPrimarySites =
+        mapToList(primarySites, x -> createProgramPrimarySite(p, x).get());
+    List<ProgramInstitution> programInstitutions =
+        mapToList(institutions, x -> createProgramInstitution(p, x).get());
+    List<ProgramCountry> programCountries =
+        mapToList(countries, x -> createProgramCountry(p, x).get());
+
+    programCancerRepository.saveAll(programCancers);
+    programPrimarySiteRepository.saveAll(programPrimarySites);
+    programInstitutionRepository.saveAll(programInstitutions);
+    programCountryRepository.saveAll(programCountries);
+
+    return programEntity;
+  }
+
+  private List<InstitutionEntity> filterAndAddInstitutions(@NonNull List<String> names) {
+    val duplicates = findDuplicates(names);
+    if (!duplicates.isEmpty()) {
+      throw Status.INVALID_ARGUMENT
+          .augmentDescription(
+              format("Please remove duplicate institutions %s in the list.", duplicates))
+          .asRuntimeException();
+    }
+    List<InstitutionEntity> institutions = institutionRepository.findAllByNameIn(names);
+    val existing =
+        institutions.stream().map(InstitutionEntity::getName).collect(toUnmodifiableSet());
+    names.stream()
+        .filter(i -> !existing.contains(i))
+        .forEach(
+            i -> {
+              val newInstitute = new InstitutionEntity();
+              newInstitute.setName(i);
+              institutionRepository.save(newInstitute);
+            });
+    institutions = institutionRepository.findAllByNameIn(names);
+    if (institutions.size() != names.size()) {
+      throw new IllegalStateException("Was unable to add new institutions"); // Final Sanity Check
+    }
+    return institutions;
+  }
+
+  public ProgramEntity updateProgram(
+      @NonNull ProgramEntity programToUpdate,
+      @NonNull ProgramEntity updatingProgram,
+      @NonNull List<String> cancers,
+      @NonNull List<String> primarySites,
+      @NonNull List<String> institutions,
+      @NonNull List<String> countries)
+      throws NotFoundException {
+    if (cancers.isEmpty()
+        || primarySites.isEmpty()
+        || institutions.isEmpty()
+        || countries.isEmpty()) {
+      throw Status.INVALID_ARGUMENT
+          .augmentDescription(
+              "Cannot update program. Cancer, primary site, institution, country, and region cannot be empty.")

```

Comment:
Shouldnt need to enforce region now, let's update the message.

Reply:
updated the message